[clang] Add some CodeGen tests for CWG 1xx issues

Endilll opened the pull request:
Covers CWG issues
[124](https://cplusplus.github.io/CWG/issues/124.html)
[185](https://cplusplus.github.io/CWG/issues/185.html),
[193](https://cplusplus.github.io/CWG/issues/193.html),
[199](https://cplusplus.github.io/CWG/issues/199.html).

I also looked at [190](https://cplusplus.github.io/CWG/issues/190.html), but concluded that we should try to test it via C++20 `std::is_layout_compatible` first.

I tried to group tests under `dr1xx-codegen.cpp`, but found out that CodeGen can arbitrarily reorder function definitions in LLVM module. In particular, interleaving between regular function definitions and destructor definitions present in the source might not be preserved, which messes up FileCheck directives. `CHECK-DAG` can help with that, but its interaction with `CHECK-LABEL` (lack of thereof) would require me to relax tests too much.

Endilll committed:
[clang] Add some CodeGen tests for CWG 1xx issues

Covers CWG issues
[185](https://cplusplus.github.io/CWG/issues/185.html),
[193](https://cplusplus.github.io/CWG/issues/193.html),
[199](https://cplusplus.github.io/CWG/issues/199.html).

I also looked at [190](https://cplusplus.github.io/CWG/issues/190.html), but concluded that we should try to test it via C++20 `std::is_layout_compatible` first.

I tried to group tests under `dr1xx-codegen.cpp`, but found out that CodeGen can arbitrarily reorder function definitions in LLVM module. In particular, interleaving between regular function definitions and destructor definition present in the source might not be preserved, which messes up FileCheck directives. CHECK-DAG can help with that, but its interaction with CHECK-LABEL (lack of thereof) would require me to relax the test too much.

erichkeane reviewed (approved):
This looks right to me.

Endilll committed:
Add test for CWG124

I don't think it's a duplicate of 201, and it's not officially recognized as such

Endilll committed:
Merge remote-tracking branch 'upstream/main' into cwg1xx-codegen

erichkeane commented on clang/test/CXX/drs/dr124.cpp:
```diff
@@ -0,0 +1,51 @@
+// CHECK-LABEL: arrayctor.loop:
+// CHECK:         call void @dr124::A::A()
+// CHECK:         call void @dr124::B::B(dr124::A)
+// CHECK:         call void @dr124::A::~A()
```
Yeah, i think this is doing what it says here.

Let @cor3ntin approve, but this looks right to me.

cor3ntin replied:
I'm not sure the `full_expr_fence` does anything here.
The salient part of the test is L41.

That being said, I'm happy with that

Endilll committed:
Merge remote-tracking branch 'upstream/main' into cwg1xx-codegen

Endilll committed:
Merge remote-tracking branch 'upstream/main' into cwg1xx-codegen
